### PR TITLE
22 POST/ sign up

### PR DESCRIPTION
**Pull request by DeenaAlghazali**

### CODE REVIEW

File: server/controllers/auth/signUp.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      );
+    }
+    const hashedPassword = await hash(password, 10);
+
+    const donor: any = await Donor.create({
+      name, email, password: hashedPassword, phone, address,
+    });
+
+    // eslint-disable-next-line max-len
+    await sign({ id: donor.id }, process.env.SECRET as string, (error:Error | null, token: string | undefined) => {
```
Comment (Mohammed-Hamada) [inline server/controllers/auth/signUp.ts:40]: Here, I create function to sign token, so try to pull main in main and use it in your branch

File: server/utils/validation/signUpSchema.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import Joi from 'joi';
+
+const signUpSchema = Joi.object({
+  name: Joi.string().required(),
+  email: Joi.string().max(250).min(4).email({ minDomainSegments: 2, tlds: { allow: ['com', 'net'] } })
+    .required(),
+  password: Joi.string().max(50).min(3).required(),
+  address: Joi.string().required(),
+  phone: Joi.string().required(),
+  is_admin: Joi.boolean(),
```
Comment (Mohammed-Hamada) [inline server/utils/validation/signUpSchema.ts:10]: There is no need to validate is_admin because here you just want to validate Sign Up form inputs.

**Review by Mohammed-Hamada — CHANGES_REQUESTED** [review]

File: server/controllers/auth/signUp.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      if (error) throw new CustomedError('Hash Function Error', 400);
+
+      res.status(201).cookie('ACCESS_TOKEN', token, {
+        maxAge: 900000,
+        httpOnly: true,
+      }).json({
+        message: 'Sign in successfully',
+        data: {
+          name, email, phone, address, is_admin: donor.is_admin,
+        },
```
Comment (Raghad-Mezied) [inline server/controllers/auth/signUp.ts:50]: Do you need all this data to send as a response?

File: server/controllers/auth/signUp.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    if (checkEmail) {
+      throw new CustomedError(
+        'Email is used try another one',
+        400,
+      );
+    }
+    const hashedPassword = await hash(password, 10);
+
+    const donor: any = await Donor.create({
```
Comment (Raghad-Mezied) [inline server/controllers/auth/signUp.ts:35]: Edit all any types :satisfied:

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

**Review by Mohammed-Hamada — APPROVED** [review]

File: server/tests/route.test.ts [hunk 5]
```diff
@@ -112,6 +112,47 @@ describe('POST/login', () => {
...
+      .post('/api/signUp')
+      .send({
+        name: 'mohammed',
+        email: 'mohaammed@gmail.com',
+        password: '123456789',
+        address: 'Gaza',
+        phone: '0599522660',
+      })
+      .expect(201);
+    expect(response.body.message).toBe('Sign up successfully');
```
Comment (Raghad-Mezied) [inline server/tests/route.test.ts:127]: you should check if your controller add a cookie or not

File: server/controllers/auth/signUp.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    res.status(201).cookie('ACCESS_TOKEN', token, {
+      maxAge: 900000,
+      httpOnly: true,
+    }).json({
+      message: 'Sign up successfully',
+      data: {
+        id: donor.id, name, is_admin: donor.is_admin,
+      },
+    });
+  } catch (error: any) {
```
Comment (Raghad-Mezied) [inline server/controllers/auth/signUp.ts:67]: add type for the error

File: server/controllers/auth/signUp.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+          'phone is used try another one',
+          400,
+        );
+      }
+    }
+
+    const hashedPassword = await hash(password, 10);
+
+    const donor: any = await Donor.create({
+      name, email, password: hashedPassword, phone, address,
```
Comment (Raghad-Mezied) [inline server/controllers/auth/signUp.ts:47]: Add each one in a new line, so you don't end up with a long line.

File: server/controllers/auth/signUp.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    const validationResult = await signUpSchema.validateAsync(req.body);
+    const {
+      name,
+      email,
+      password,
+      phone,
+      address,
+    } = validationResult;
+
+    const check: any = await Donor.findOne({
```
Comment (Raghad-Mezied) [inline server/controllers/auth/signUp.ts:21]: edit the type

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

**Review by farah2003 — APPROVED** [review]

**Review by HaniOlwan — APPROVED** [review]

**Pull request merged**